### PR TITLE
Make `tape.trainable_params` a list

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
Recently, the `trainable_params` property of `QuantumTape`s, which is a `set`, caused bugs based on the fact that it is not guaranteed to maintain its ordering.
In particular for larger tapes these rare instances seem to occur, which likely is the reason why it was not caught before.
As an example, consider the "unordered" set `{3, 1}`. As integer hashes are the integers themselves, a "canonical" ordering for python is according to the hash table, so that
```pycon
>>> {3, 1}
{1, 3}
```
This makes constructing small tests that fail based on `trainable_params` being a set quite hard.

**Description of the Change:**
`QuantumTape.trainable_params` is made a `list` instead, with an intermediate conversion to `set` and subsequent sorting within the attribute setter function to exclude duplicates and assure the entries are sorted. Whenever the uniqueness and ordering of the entries is guaranteed, the private attribute `_trainable_params` may be set directly, as is done in `_update_trainable_params`.

**Benefits:**
Fixes rare but apparently increasingly important edge case bugs where the operation/parameter ordering was disturbed by the loss of ordering within `trainable_params`.

**Possible Drawbacks:**
For very large tapes with many trainable parameters, a lookup will be slower.

**Related GitHub Issues:**
--

### CODE REVIEW

File: pennylane/tape/tape.py [hunk 1]
```diff
@@ -754,14 +756,17 @@ def get_operation(self, idx):
...
+        info = self._par_info[t_idx]
 
         # get the corresponding operation
-        op = self._par_info[t_idx]["op"]
+        op = info["op"]
 
         # get the corresponding operation parameter index
         # (that is, index of the parameter within the operation)
-        p_idx = self._par_info[t_idx]["p_idx"]
+        p_idx = info["p_idx"]
```
Comment (josh146) [inline pennylane/tape/tape.py:769]: a nice micro-optimization :)

**Review by josh146 — APPROVED** [review]
thanks for fixing this @dwierichs!

File: pennylane/tape/tape.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/tape/tape.py:769]: :)

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -37,6 +37,11 @@
 
 <h3>Breaking changes</h3>
 
+* `QuantumTape.trainable_params` now is a list instead of a set. This
+  means that `tape.trainable_params` will return a list unlike before,
+  but setting the `trainable_params` with a set works exactly as before.
+  [(#1xxx)](https://github.com/PennyLaneAI/pennylane/pull/1xxx)
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/releases/changelog-dev.md:43]: @dwierichs I think you forgot this one 😆 I'll incorporate it into #1807, I'm currently resolving changelog conflicts over there

[the remark this answers is not in the document]
Comment (dwierichs) [reply doc/releases/changelog-dev.md:43]: Yeah, sorry, I forgot the second occurrence of the PR number in there :sweat_smile: Maria already pointed this out to me as well :D
Thanks for taking care of it! :)

**Pull request merged**